### PR TITLE
timer: add sigev_thread flag support for timer_create in zephyr posix

### PR DESCRIPTION
**Pull request by arun-mani-tech**
this commit adds support for the sigev_thread flag in the timer_create function of the zephyr posix layer. with this enhancement, user can now specify the sigev_thread flag when creating the timer, allowing the timer notofication to be delivered to a dedicated thread rather than a signal or none. the new thread can be configured with specific attributes, such as stacksize and priority, to tailer its behavior for different use cases

sigev_thread flag works on per timer per thread method.

### CODE REVIEW

**Comment by ssekar15** [comment]
use pthread instead of k_thread to avoid addnl changes.

File: lib/posix/timer.c [hunk 1]
```diff
@@ -75,6 +109,31 @@ int timer_create(clockid_t clockid, struct sigevent *evp, timer_t *timerid)
...
+	} else if (evp->sigev_notify == SIGEV_THREAD) {
+		int ret;
+		size_t total_size;
+
+		timer->sigev_notify = SIGEV_THREAD;
+		__alignof__(Z_KERNEL_STACK_OBJ_ALIGN);
+		total_size = Z_KERNEL_STACK_SIZE_ADJUST(CONFIG_TIMER_STACK_SIZE);
+		timer->stack = (struct z_thread_stack_element *)k_malloc(total_size);
+		if (!timer->stack)
+			return -EIO;
```
Comment (parthitce) [inline lib/posix/timer.c:121]: ENOMEM?

File: lib/posix/timer.c [hunk 2]
```diff
@@ -198,6 +257,18 @@ int timer_delete(timer_t timerid)
 		k_timer_stop(&timer->ztimer);
 	}
 
+	if (timer->sigev_notify == SIGEV_THREAD) {
+		pthread_cancel(timer->thread);
```
Comment (parthitce) [inline lib/posix/timer.c:261]: should we `pthread_join` ?

File: lib/posix/timer.c [hunk 3]
```diff
@@ -198,6 +257,18 @@ int timer_delete(timer_t timerid)
...
+
+		if (timer->stack != NULL) {
+			k_free(timer->stack);
+			timer->stack = NULL;
+		}
+
```
Comment (parthitce) [inline lib/posix/timer.c:267]: `pthread_attr_destroy` ?

**Review by parthitce — COMMENTED** [review]
Fix the compliance checks.

File: lib/posix/timer.c [hunk 4]
```diff
@@ -75,6 +109,31 @@ int timer_create(clockid_t clockid, struct sigevent *evp, timer_t *timerid)
 
 	if (evp->sigev_notify == SIGEV_NONE) {
 		k_timer_init(&timer->ztimer, NULL, NULL);
+	} else if (evp->sigev_notify == SIGEV_THREAD) {
+		int ret;
+		size_t total_size;
+
+		timer->sigev_notify = SIGEV_THREAD;
+		__alignof__(Z_KERNEL_STACK_OBJ_ALIGN);
```
Comment (parthitce) [inline lib/posix/timer.c:117]: what does this do?

File: lib/posix/timer.c (hunk 2 above)
Comment (arun-mani-tech) [reply lib/posix/timer.c:261]: if thread takes long time to end, it become leads to hold here. but we need to stop the thread when timer is deleted right?
so thats why i use thread_cancel here

File: lib/posix/timer.c [hunk 5]
```diff
@@ -75,6 +109,34 @@ int timer_create(clockid_t clockid, struct sigevent *evp, timer_t *timerid)
...
+		int ret;
+		size_t total_size;
+
+		timer->sigev_notify = SIGEV_THREAD;
+		total_size = Z_KERNEL_STACK_SIZE_ADJUST(CONFIG_TIMER_STACK_SIZE);
+		timer->stack = (struct z_thread_stack_element *)k_malloc(total_size);
+		if (!timer->stack)
+			return -ENOMEM;
+		else
+			timer->stack = (struct z_thread_stack_element *)(((uintptr_t)timer->stack +
```
Comment (parthitce) [inline lib/posix/timer.c:122]: mind explaining what's happening here?

File: lib/posix/timer.c [hunk 6]
```diff
@@ -39,6 +46,29 @@ static void zephyr_timer_wrapper(struct k_timer *ztimer)
...
+static void *zephyr_thread_wrapper(void *arg)
+{
+	struct timer_obj *timer = (struct timer_obj *)arg;
+
+	while (1) {
+		if (timer->reload == 0U) {
+			timer->status = NOT_ACTIVE;
+		}
+
+		pthread_cond_wait(&timer->cond, &timer->lock);
```
Comment (ssekar15) [inline lib/posix/timer.c:58]: use of pthread_cond_wait is wrong.. read man page

File: lib/posix/timer.c [hunk 7]
```diff
@@ -75,6 +106,33 @@ int timer_create(clockid_t clockid, struct sigevent *evp, timer_t *timerid)
...
+		timer->sigev_notify = SIGEV_THREAD;
+		total_size = Z_KERNEL_STACK_SIZE_ADJUST(CONFIG_TIMER_STACK_SIZE);
+		timer->stack = (struct z_thread_stack_element *)
+				k_aligned_alloc(Z_KERNEL_STACK_OBJ_ALIGN, total_size);
+		if (!timer->stack) {
+			k_mem_slab_free(&posix_timer_slab, (void *) &timer);
+			return -ENOMEM;
+		}
+
+		pthread_cond_init(&timer->cond, NULL);
```
Comment (ssekar15) [inline lib/posix/timer.c:122]: where is mutex initialization ?

File: lib/posix/timer.c (hunk 6 above)
Comment (ssekar15) [reply lib/posix/timer.c:58]: I'm not convinced using pthread_cond with mutex, why not simple binary semaphore !

**Comment by ssekar15** [comment]
look fine.. create PR in upstream

**Comment by DineshDK03** [comment]
The following west manifest projects have been modified in this Pull Request:

| Name | Old Revision | New Revision | Diff |
| ---- | ------------ | ------------ |------|
| hal_espressif | https://github.com/zephyrproject-rtos/hal_espressif/commit/9531c1c17873aa4935f85864ebafd20c02062a8f | https://github.com/zephyrproject-rtos/hal_espressif/commit/ae1bd648a1ac701672c46b6ff4eadfa716937ce3 (zephyr) | [zephyrproject-rtos/hal_espressif@9531c1c1..ae1bd648](https://github.com/zephyrproject-rtos/hal_espressif/compare/9531c1c17873aa4935f85864ebafd20c02062a8f..ae1bd648a1ac701672c46b6ff4eadfa716937ce3) |
| hal_nordic | https://github.com/zephyrproject-rtos/hal_nordic/commit/a1c3e0fbaafda091139b8744becd4853ada2f747 | https://github.com/zephyrproject-rtos/hal_nordic/commit/cf6e9fc5f7c2c98df26f2a4227a95df9a50823e7 (master) | [zephyrproject-rtos/hal_nordic@a1c3e0fb..cf6e9fc5](https://github.com/zephyrproject-rtos/hal_nordic/compare/a1c3e0fbaafda091139b8744becd4853ada2f747..cf6e9fc5f7c2c98df26f2a4227a95df9a50823e7) |
| hal_nuvoton | https://github.com/zephyrproject-rtos/hal_nuvoton/commit/8a2b5de1670b59fcacd20d7495cb1c0f26fbe7bd | https://github.com/zephyrproject-rtos/hal_nuvoton/commit/3e0a4c4d3328b2f72b164219add19d5308b53cb5 (master) | [zephyrproject-rtos/hal_nuvoton@8a2b5de1..3e0a4c4d](https://github.com/zephyrproject-rtos/hal_nuvoton/compare/8a2b5de1670b59fcacd20d7495cb1c0f26fbe7bd..3e0a4c4d3328b2f72b164219add19d5308b53cb5) |
| hal_nxp | https://github.com/zephyrproject-rtos/hal_nxp/commit/e0116b8eae46c2d90f2c1b23baef73bb082d00bd | https://github.com/zephyrproject-rtos/hal_nxp/commit/3aed18634b19102cc0f9e56fcff88bd04edba6ec | [zephyrproject-rtos/hal_nxp@e0116b8e..3aed1863](https://github.com/zephyrproject-rtos/hal_nxp/compare/e0116b8eae46c2d90f2c1b23baef73bb082d00bd..3aed18634b19102cc0f9e56fcff88bd04edba6ec) |
| hal_silabs | https://github.com/zephyrproject-rtos/hal_silabs/commit/a143f03e846eb1b7b3135f3c8192820ce1b6d9c4 | https://github.com/zephyrproject-rtos/hal_silabs/commit/5fbe78d3676b5500585f8e4534a8e56e7dae74a8 | [zephyrproject-rtos/hal_silabs@a143f03e..5fbe78d3](https://github.com/zephyrproject-rtos/hal_silabs/compare/a143f03e846eb1b7b3135f3c8192820ce1b6d9c4..5fbe78d3676b5500585f8e4534a8e56e7dae74a8) |
| hal_stm32 | https://github.com/zephyrproject-rtos/hal_stm32/commit/c865374fc83d93416c0f380e6310368ff55d6ce2 | https://github.com/zephyrproject-rtos/hal_stm32/commit/1bc72c299d0365c0ee2575a97918b22df0899e10 | [zephyrproject-rtos/hal_stm32@c865374f..1bc72c29](https://github.com/zephyrproject-rtos/hal_stm32/compare/c865374fc83d93416c0f380e6310368ff55d6ce2..1bc72c299d0365c0ee2575a97918b22df0899e10) |
| lvgl | https://github.com/zephyrproject-rtos/lvgl/commit/7102083f626cda09e5792420ea60af0525cce9ae | https://github.com/zephyrproject-rtos/lvgl/commit/5da257f782a8f9c6e265bdc60ebc2a93fdee24de (zephyr) | [zephyrproject-rtos/lvgl@7102083f..5da257f7](https://github.com/zephyrproject-rtos/lvgl/compare/7102083f626cda09e5792420ea60af0525cce9ae..5da257f782a8f9c6e265bdc60ebc2a93fdee24de) |
| mcuboot | https://github.com/zephyrproject-rtos/mcuboot/commit/74c4d1c52fd51d07904b27a7aa9b2303e896a4e3 | https://github.com/zephyrproject-rtos/mcuboot/commit/76d19b3b8885ea7ae25a6f4f5d8501f7ec646447 (main,upstream-sync) | [zephyrproject-rtos/mcuboot@74c4d1c5..76d19b3b](https://github.com/zephyrproject-rtos/mcuboot/compare/74c4d1c52fd51d07904b27a7aa9b2303e896a4e3..76d19b3b8885ea7ae25a6f4f5d8501f7ec646447) |
| nrf_hw_models | https://github.com/zephyrproject-rtos/nrf_hw_models/commit/33b999d50093da88cf935383da2bdac70ffd9456 | https://github.com/zephyrproject-rtos/nrf_hw_models/commit/cee41373eb0bb720a370586de1ee5eb693e62c95 | [zephyrproject-rtos/nrf_hw_models@33b999d5..cee41373](https://github.com/zephyrproject-rtos/nrf_hw_models/compare/33b999d50093da88cf935383da2bdac70ffd9456..cee41373eb0bb720a370586de1ee5eb693e62c95) |
| open-amp | https://github.com/zephyrproject-rtos/open-amp/commit/c904a01d4a882bcbb39987e0e2ce5308f49ac7ad | https://github.com/zephyrproject-rtos/open-amp/commit/42b7c577714b8f22ce82a901e19c1814af4609a8 (master) | [zephyrproject-rtos/open-amp@c904a01d..42b7c577](https://github.com/zephyrproject-rtos/open-amp/compare/c904a01d4a882bcbb39987e0e2ce5308f49ac7ad..42b7c577714b8f22ce82a901e19c1814af4609a8) |
| openthread | https://github.com/zephyrproject-rtos/openthread/commit/37fb77098982d17555dd7a3f58832714bb9df56e | https://github.com/zephyrproject-rtos/openthread/commit/f7690fe7e9d638341921808cba6a3e695ec0131e (main) | [zephyrproject-rtos/openthread@37fb7709..f7690fe7](https://github.com/zephyrproject-rtos/openthread/compare/37fb77098982d17555dd7a3f58832714bb9df56e..f7690fe7e9d638341921808cba6a3e695ec0131e) |
| segger | https://github.com/zephyrproject-rtos/segger/commit/4bfaf28a11c3e5ec29badac744fab6d2f342749e | https://github.com/zephyrproject-rtos/segger/commit/5792675a2470d0f3857de1e77bff57b38c28de3b (master) | [zephyrproject-rtos/segger@4bfaf28a..5792675a](https://github.com/zephyrproject-rtos/segger/compare/4bfaf28a11c3e5ec29badac744fab6d2f342749e..5792675a2470d0f3857de1e77bff57b38c28de3b) |

*Note: This message is automatically posted and updated by the Manifest GitHub Action.*

**Pull request closed without merging**